### PR TITLE
fix(sql): bug in SAMPLE BY queries after a query cache hit

### PR DESCRIPTION
**Pull request by jerrinot**
This bug manifests as QuestDB error complaining about an invalid sampling interval even when the interval is valid. 

Root cause: `FillRangeRecordCursorFactory` would store a reference to a mutable CharSequence token representing the sampling interval. This error could occur if the interval changed between factory instantiation and its usage. I believe it could even crash QuestDB in the worst case, but I haven't managed to reproduce this. 

This PR fixes this by eagerly parsing the interval CharSequence and storing just immutable values.

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 24 / 26 (92.31%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|5|5|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 32 / 34 (94.12%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|13|15|86.67%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java [hunk 1]
```diff
@@ -125,7 +132,9 @@ public void toPlan(PlanSink sink) {
         if (fromFunc != TimestampConstant.NULL || toFunc != TimestampConstant.NULL) {
             sink.attr("range").val('(').val(fromFunc).val(',').val(toFunc).val(')');
         }
-        sink.attr("stride").val('\'').val(stride).val('\'');
+        // todo: this might be incorrect, since I dunno if concatenation of a long and char works as expected
```
[the remark this answers is not in the document]
Comment (jerrinot) [reply core/src/main/java/io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java:135]: good eyes, thanks!

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

File: core/src/main/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java [hunk 2]
```diff
@@ -33,21 +33,25 @@
...
+     * Find the end of the interval token in the input string. The interval token is expected to be a number followed by
+     * a single letter qualifier.
      *
-     * @param cs       the key
+     * @param cs       input string
      * @param position position in SQL text to report error against
-     * @return instance of appropriate TimestampSampler
-     * @throws SqlException when input string is invalid
+     * @return index of the first character after the interval token
+     * @throws SqlException
```
Comment (bluestreak01) [inline core/src/main/java/io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java:42]: nit: this is a new javadoc warning, undocumented exception

File: core/src/test/java/io/questdb/test/griffin/engine/groupby/TimestampSamplerFactoryTest.java [hunk 3]
```diff
@@ -180,6 +215,16 @@ public void testUnsupportedQualifier2() {
         assertFailure(132, "unsupported interval qualifier", "21K", 130);
     }
 
+    private static void assertFindIntervalEndIndexFailure(int expectedPosition, CharSequence expectedMessage, CharSequence sampleBy, int position) {
```
Comment (bluestreak01) [inline core/src/test/java/io/questdb/test/griffin/engine/groupby/TimestampSamplerFactoryTest.java:218]: nit: can we use different numbers for "position" parameter, right now it is hardcoded to the same value of 42?

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 37 / 37 (100.00%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/TimestampSamplerFactory.java|18|18|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/FillRangeRecordCursorFactory.java|13|13|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|6|6|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**